//
// gRPC definition for the OBD Service
//
//	Copyright Laurent Carré - Sterwen Technology 2019-2020
//	Version 0.9	28/01/2020
//
syntax="proto3";

//  =================================================================
//  Messages definition
//  =================================================================
//

message Start_OBD {
	//
	// Message needed to start connection with OBD dongle
	//
	string MAC=1;	// MAC address of the dongle
	
	}
	

message OBD_CMD_value {
	string cmd=1;
	enum value_type {
		Quantity=0;
		Other=1;
		}
	value_type type =2;
	oneof value {
		float f=4;
		string s=5;
		}
	string unit=10;
}
	
message OBD_Result {
		bool connected=1;
		bool engine_on=2;
		string error=3;
		string obd_time=4;
		repeated OBD_CMD_value values=5; // only if connected + engine on	
}

message OBD_cmd {
	enum RequestResult {
		Default = 0;
		Specific=1;
		}
	RequestResult request=1;	
	string commands=2; // only if request == specific (comma separated values)
	string rules=4; // JSON encoded
}

message OBD_StatusRequest {
	string request=1;  // min, vehicle_cmds, active_cmds, actual_cmds
}

message OBD_status {
	bool connected =1;
	bool engine_on =2;
	bool autoconnect=3;
	string state = 7; // if connected
	string protocol=4; // if connected
	string MAC = 5; // if autoconnect = True
	string commands = 6; // (comma separated values)
}
	


//  ==============================================================
//	Service definition
//  ===============================================================

service OBD_Service {
	rpc Status(OBD_StatusRequest) returns (OBD_status);
	rpc Connect(Start_OBD) returns(OBD_Result);
	rpc Read(OBD_cmd) returns (stream OBD_Result);
	rpc Stop(OBD_cmd) returns (OBD_Result);	
	}